server/src/main/java/org/candlepin/model/OwnerProductCurator.java

```diff
@@ -0,0 +1,400 @@
+    }
+
+    @Transactional
+    public boolean isProductMappedToOwner(Product product, Owner owner) {
+        String jpql = "SELECT count(op) FROM OwnerProduct op " +
+            "WHERE op.owner.id = :owner_id AND op.product.uuid = :product_uuid";
+
+        long count = (Long) this.getEntityManager()
+            .createQuery(jpql)
+            .setParameter("owner_id", owner.getId())
+            .setParameter("product_uuid", product.getUuid())
+            .getSingleResult();
+
+        return count > 0;
+    }
+
+    @Transactional
+    public boolean mapProductToOwner(Product product, Owner owner) {
+        if (!this.isProductMappedToOwner(product, owner)) {
+            this.create(new OwnerProduct(owner, product));
+
+            return true;
+        }
+
+        return false;
+    }
+
+    @Transactional
+    public int mapProductToOwners(Product product, Owner... owners) {
+        int count = 0;
+
+        for (Owner owner : owners) {
+            if (this.mapProductToOwner(product, owner)) {
+                ++count;
+            }
+        }
+
+        return count;
+    }
+
+    @Transactional
+    public int mapOwnerToProducts(Owner owner, Product... products) {
+        int count = 0;
+
+        for (Product product : products) {
+            if (this.mapProductToOwner(product, owner)) {
+                ++count;
+            }
+        }
+
+        return count;
+    }
+
+    @Transactional
+    public boolean removeOwnerFromProduct(Product product, Owner owner) {
+        String jpql = "DELETE FROM OwnerProduct op " +
+            "WHERE op.product.uuid = :product_uuid AND op.owner.id = :owner_id";
+
+        int rows = this.getEntityManager()
+            .createQuery(jpql)
+            .setParameter("owner_id", owner.getId())
+            .setParameter("product_uuid", product.getUuid())
+            .executeUpdate();
+
+        return rows > 0;
+    }
+
+    @Transactional
+    public int clearOwnersForProduct(Product product) {
+        String jpql = "DELETE FROM OwnerProduct op " +
+            "WHERE op.product.uuid = :product_uuid";
+
+        return this.getEntityManager()
+            .createQuery(jpql)
+            .setParameter("product_uuid", product.getUuid())
+            .executeUpdate();
+    }
+
+    @Transactional
+    public int clearProductsForOwner(Owner owner) {
+        String jpql = "DELETE FROM OwnerProduct op " +
+            "WHERE op.owner.id = :owner_id";
+
+        return this.getEntityManager()
+            .createQuery(jpql)
+            .setParameter("owner_id", owner.getId())
+            .executeUpdate();
+    }
+
+    /**
+     * Updates the product references currently pointing to the original product to instead point to
+     * the updated product for the specified owners.
+     * <p/></p>
+     * <strong>Warning:</strong> Hibernate does not gracefully handle situations where the data
+     * backing an entity changes via direct SQL or other outside influence. While, logically, a
```

Comment:
Just a small note about this. I think this behavior is caused by cascading. The same goes for OwnerContentCurator.testRemoveOwnerContentReferences where you have to evict and then find instead of just refresh. Refresh there will first try to cascade the refresh to collection which causes exceptions because the elements no longer exist in the database.

Yeah I dont think there is anything to be done about this. Eitehr clear() or evict() seems the most bulletproof response in these situations...